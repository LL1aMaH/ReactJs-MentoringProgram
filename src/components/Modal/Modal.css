.layout {
  display: flex;
  justify-content: center;
  align-items: center;
  position: fixed;
  top: 0;
  left: 0;
  width: 100vw;
  height: 100vh;
  padding: 4rem;
  overflow-y: auto;
  overflow-x: hidden;
  background-color: rgba(36, 43, 54, 0.7);
  z-index: 10;
  animation: softAppear 200ms;
}

.content {
  margin: auto 0;
  max-width: 100%;
}

.scrollLock {
  overflow: hidden;
}

@keyframes softAppear {
  from {
    opacity: 0;
  }

  to {
    opacity: 1;
  }
}
