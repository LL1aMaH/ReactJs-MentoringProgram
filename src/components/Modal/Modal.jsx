import React, { memo, useEffect, useCallback } from 'react';
import { createPortal } from 'react-dom';

import FocusLock from 'react-focus-lock';

import styles from '../../../styles/Modal.module.scss'

export const Modal = memo(function Modal(props){
  let {
    open,
    onClose,
    children,
    container,
    disableCloseOnBackground = false,
    contentClassName: customContentClassName,
    backgroundClassName: customBackgroundClassName,
  } = props;

  useEffect(() => 
  {
    container = document.body
  },[])

  const layoutClassName = `${styles.layout} ${customBackgroundClassName}`;

  const contentClassName = `${styles.content}  ${customContentClassName}`;

  const handleBackgroundClick = useCallback(
    (e) => {
      const { target, currentTarget } = e;
      if (!disableCloseOnBackground && onClose && target === currentTarget) {
        onClose(e);
      }
    },
    [onClose, disableCloseOnBackground],
  );

  useEffect(() => {
    const root = document.body;

    if (open) {
      root.classList.add(styles.scrollLock);
    } else {
      root.classList.remove(styles.scrollLock);
    }

    return () => {
      root.classList.remove(styles.scrollLock);
    };
  }, [open]);



  const modal = open && (
    <div
      className={layoutClassName}
      onClick={handleBackgroundClick}
      data-testid="modal"
      role="presentation"
    >
      <FocusLock className={contentClassName}>{children}</FocusLock>
    </div>
  );

  return createPortal(modal, container);
});
