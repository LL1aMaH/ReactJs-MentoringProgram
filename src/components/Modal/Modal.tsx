/* eslint-disable react/prop-types */
import React, { memo, useEffect, useCallback, ReactNode } from 'react';
import { createPortal } from 'react-dom';

import FocusLock from 'react-focus-lock';

import styles from './Modal.css';

export type TModalProps = {
  open: boolean;
  children: ReactNode;
  onClose?: (e: object) => void;
  container?: Element;
  disableCloseOnBackground?: boolean;
  backgroundClassName?: string;
  contentClassName?: string;
};

export const Modal = memo(function Modal(props: TModalProps): JSX.Element {
  const {
    open,
    onClose,
    children,
    container = document.body,
    disableCloseOnBackground = false,
    contentClassName: customContentClassName,
    backgroundClassName: customBackgroundClassName,
  } = props;

  const layoutClassName = `${styles.layout} ${customBackgroundClassName}`;

  const contentClassName = `${styles.content}  ${customContentClassName}`;

  const handleBackgroundClick = useCallback(
    (e) => {
      const { target, currentTarget } = e;
      if (!disableCloseOnBackground && onClose && target === currentTarget) {
        onClose(e);
      }
    },
    [onClose, disableCloseOnBackground],
  );

  useEffect(() => {
    const root = document.body;

    if (open) {
      root.classList.add(styles.scrollLock);
    } else {
      root.classList.remove(styles.scrollLock);
    }

    return () => {
      root.classList.remove(styles.scrollLock);
    };
  }, [open]);

  const modal = open && (
    <div
      className={layoutClassName}
      onClick={handleBackgroundClick}
      data-testid="modal"
      role="presentation"
    >
      <FocusLock className={contentClassName}>{children}</FocusLock>
    </div>
  );

  return createPortal(modal, container);
});
