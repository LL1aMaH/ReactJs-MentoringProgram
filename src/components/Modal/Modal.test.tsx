import React, { ReactNode } from 'react';
import { cleanup, fireEvent, render } from '@testing-library/react';
import { Modal } from './Modal';

jest.mock(
  'react-focus-lock',
  () => ({ children }: { children: ReactNode }) => <div id="Focus lock">{children}</div>, // eslint-disable-line react/display-name
);

describe('Modal', () => {
  const getComponent = (props: { onClose?: () => void; disableCloseOnBackground?: boolean }) =>
    render(
      <Modal open {...props}>
        <div data-testid="content">content</div>
      </Modal>,
    );

  afterEach(cleanup);

  it('should render correctly', () => {
    const { getByTestId } = getComponent({});

    expect(getByTestId('modal')).toMatchSnapshot();
  });

  it('should call onClose on click background', () => {
    const onClose = jest.fn();
    const { getByTestId } = getComponent({ onClose });

    fireEvent.click(getByTestId('modal'));

    expect(onClose).toBeCalledTimes(1);
  });

  it('should not call onClose on click content', () => {
    const onClose = jest.fn();
    const { getByTestId } = getComponent({ onClose });

    fireEvent.click(getByTestId('content'));

    expect(onClose).not.toBeCalled();
  });

  it('should not call onClose on click background when disableCloseOnBackground is set to true', () => {
    const onClose = jest.fn();
    const { getByTestId } = getComponent({ onClose, disableCloseOnBackground: true });

    fireEvent.click(getByTestId('modal'));

    expect(onClose).not.toBeCalled();
  });

  it('should not call onClose on press any other button', () => {
    const onClose = jest.fn();
    const { getByTestId } = getComponent({ onClose });

    fireEvent.keyDown(getByTestId('modal'), { keyCode: 20 });

    expect(onClose).not.toBeCalled();
  });

  it('should lock body scroll when open is true', () => {
    const { baseElement, unmount } = getComponent({});

    expect(baseElement.classList.contains('scrollLock')).toBeTruthy();

    unmount();

    expect(baseElement.classList.contains('scrollLock')).toBeFalsy();
  });
});
