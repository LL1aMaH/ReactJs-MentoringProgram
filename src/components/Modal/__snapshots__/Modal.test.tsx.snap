// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal should render correctly 1`] = `
<div
  class="layout"
  data-testid="modal"
  role="presentation"
>
  <div
    id="Focus lock"
  >
    <div
      data-testid="content"
    >
      content
    </div>
  </div>
</div>
`;
